### PR TITLE
[1.2-maint] Fix test_size_on_disk_accurate for large st_blksize, fixes #7250

### LINKED ISSUE #7250: test_size_on_disk_accurate on 1.2.3 breaks when run on ZFS
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Bug/Regression

#### Operating system (distribution) and version.

NixOS Unstable

#### Hardware / network configuration, and filesystems used.

ZFS Root

#### How much data is handled by borg?

n/a

#### Full borg commandline that lead to the problem (leave away excludes and passwords)

n/a

#### Can you reproduce the problem? If so, describe how. If not, describe troubleshooting steps you took before opening the issue.

Running the test suite on ZFS breaks on `test_size_on_disk_accurate` since https://github.com/borgbackup/borg/commit/8ddfd94f0904449175151faed67f9640710b6454.

ZFS has a commit interval of 5 seconds, so a manual flush might be required for the size to be reflected instantly.

#### Include any warning/errors/backtraces from the system logs

```python
_______________ HashIndexSizeTestCase.test_size_on_disk_accurate _______________
[gw42] linux -- Python 3.10.9 /nix/store/ar76b3831ydkmfhf0r9v3wbma5i61iz2-python3-3.10.9/bin/python3.10
self = <borg.testsuite.hashindex.HashIndexSizeTestCase testMethod=test_size_on_disk_accurate>

    def test_size_on_disk_accurate(self):
        idx = ChunkIndex()
        for i in range(1234):
            idx[H(i)] = i, i**2, i**3
        with tempfile.NamedTemporaryFile() as file:
            idx.write(file)
            size = os.path.getsize(file.fileno())
>       assert idx.size() == size
E       assert 90350 == 0
E        +  where 90350 = <built-in method size of borg.hashindex.ChunkIndex object at 0xffffeea51b30>()
E        +    where <built-in method size of borg.hashindex.ChunkIndex object at 0xffffeea51b30> = <borg.hashindex.ChunkIndex object at 0xffffeea51b30>.size

/nix/store/0ppx77a4xk6nij2p1klgsnksmrfnx9lj-borgbackup-1.2.3/lib/python3.10/site-packages/borg/testsuite/hashindex.py:202: AssertionError
```

### PR DESCRIPTION
**Pull request by puetzk**
python's io.BufferedWriter sizes its buffer based on st_blksize

If the write fits in this buffer, then it's possible none (or only some) of the data from idx.write() has been written through to fileno()

Note:1.2-maint counterpart of #7251.

I see that the contribution guidelines ask for PRs only against master (leaving backports up to you maintainers) but in this case I want to publicly post a commit based on 1.2-maint to this repo, so that @mweinelt can have the nixpkgs derivation cherry-pick that commit for testing. I think we understand what's going on, but the failure is being seen on nixos autobuilders that use ZFS for $TEMP, and I don't personally have anything with that configuration, or direct access to those build machines.

Fixes #7250

### CODE REVIEW

File: src/borg/testsuite/hashindex.py [hunk 1]
```diff
@@ -198,6 +198,7 @@ def test_size_on_disk_accurate(self):
             idx[H(i)] = i, i**2, i**3
         with tempfile.NamedTemporaryFile() as file:
             idx.write(file)
+            file.flush()
             size = os.path.getsize(file.fileno())
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/hashindex.py:202]: that pattern (write + getsize) is seen at quite some places in the code. did you only see it failing here?

Comment (puetzk) [reply src/borg/testsuite/hashindex.py:202]: Yeah, this was the only failure @mweinelt said the builders were hitting, and he said they passed with this patch: https://github.com/NixOS/nixpkgs/pull/207631#issuecomment-1375094224

Could be the other writes are just larger though (enough to be bigger than his st_blksize) . Though I think the key is that I don't see any others using getsize.*fileno - the others still seem to be writing to the filepath, and when you pass `str` to IndexBase.write it will be open/write/close/getsize, which will work (close causes a flush). See analysis at I posted in https://github.com/borgbackup/borg/issues/7250#issuecomment-1375100187

Comment (puetzk) [reply src/borg/testsuite/hashindex.py:202]: Reviewing all usage of `os.path.getsize()` (I see 9 of them), I don't find any more that seem problematic in 1.2-maint:

- Needs explicit `flush()` (added by this PR)
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/testsuite/hashindex.py#L200-L202

- using the `idx.write(str)`, so flushed by the implicit close following [`with open(path,...):` inside  `IndexBase.write`](https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/hashindex.pyx#L117-L119)
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/testsuite/hashindex.py#L101-L102
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/testsuite/hashindex.py#L106-L107
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/testsuite/hashindex.py#L111-L112

- `getsize` is outside of the [with open(fname, 'wb') as fd:](https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/testsuite/chunker_pytest.py#L50) block
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/testsuite/chunker_pytest.py#L50

- No preceding open/write, AFAICT checking an pre-existing file unless there their caller is possibly passing a filename they just wrote to and still have open. But it doesn't look like it to me (not very experienced with borg's code, though).
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/repository.py#L499
  https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/repository.py#L1482
  https://github.com/borgbackup/borg/blob/1ac19b034bdb2d714a58fb01cbb8a8fda891dd90/src/borg/repository.py#L1485
 https://github.com/borgbackup/borg/blob/1ac19b034bdb2d714a58fb01cbb8a8fda891dd90/src/borg/repository.py#L1533
  Though this last one attempts to open and write into `filename` if (and after) getsize indicates that the existing file is too small. But no race there.

Comment (ThomasWaldmann) [reply src/borg/testsuite/hashindex.py:202]: Thanks for checking these.

I also had another look and found that other tests simply avoid this issue by using `unopened_tempfile` context manager.

So I guess the easiest solution is to also use that CM here and not create troubles by trying to work with the open file.

See e.g. `test_resize` in `testsuite/hashindex.py`.

Comment (puetzk) [reply src/borg/testsuite/hashindex.py:202]: Sure, that certainly avoids the problem (since the file handle has been closed). It changes the coverage of this test (to the [`else:` path of `IndexBase.write`](https://github.com/borgbackup/borg/blob/537a814e53e20013a041faa7192da005f137cf5b/src/borg/hashindex.pyx#L120-L121)), but that's what it was covering before 8ddfd94f0904449175151faed67f9640710b6454. And it just avoids the windows/unix difference in tempfile.NamedTemporaryFile too (that you cited as the reason for that commit).

It seems that passing fileno to os.path.getsize() is undocumented functionality:
> https://docs.python.org/3/library/os.path.html#module-os.path says 
> Changed in version 3.6: Accepts a [path-like object](https://docs.python.org/3/glossary.html#term-path-like-object).

but fileno is a bare fd, which is not a [path-like object](https://docs.python.org/3/library/os.html#os.PathLike) (no `__fspath__`).

It clearly works in practice, seemingly because the [cpython implentation defers to `os.stat`](https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Lib/genericpath.py#L50), and a  [subset of the function in os.* accept fds and transparently substitute](https://docs.python.org/3/library/os.html#files-and-directories) (e.g. `fstat` in this case): see . `fstat`'s been around basically forever (Version 7 AT&T UNIX, I think?), so `os.stat in os.supports_fd` is true for me, and probably for just about everyone. But still, using it with getsize accepting it was undocumented, and stat accepting it is platform-dependent.

So, long story short, `unopened_tempfile` it sounds good to me :-)

Comment (ThomasWaldmann) [reply src/borg/testsuite/hashindex.py]: ok, I'll merge it. thanks for your investigations!
